W-14792370 update feature implementation gr (#487)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-configure-features-http-request.adoc

= Performing an HTTP Call
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> To interact with external services, the policy must use HTTP calls. You can perform HTTP calls in the `[entrypoint]` function, the request filter, or the response filter. To perform an HTTP call, inject the `HttpClient` as follows:

Reviewer: I believe calls can not be performed in the entrypoint, only during the request and response handlers.

What happens in the entrypoint and on the `on_request` and `on_response` is the injection of the HTTP client, so it can be later used in the code.

* Injection for `[entrypoint]` function:
+
[source,Rust]
----
#[entrypoint]
async fn configure(
    launcher: Launcher,
    Configuration(configuration): Configuration,
    client: HttpClient,
) -> Result<(), LaunchError> {
----

> * Injection for `on_request` or `on_response` wrapped functions functions:

Reviewer: ```suggestion
* Injection for `on_request` or `on_response` wrapped functions:
```

+
[source,Rust]
----
// Handler with the logic for incoming requests
​​async fn request_filter(state: RequestState, conf: &Config, client: HttpClient) {
 ...
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(bytes): Configuration) -> Result<()> {
    let config: Config = serde_json::from_slice(&bytes).unwrap();

    // For incoming request events, using the on_request wrapper, inject the client and pass it to the request_filter handler
    launcher
        .launch(on_request(|request, client| {
            request_filter(request, &config, client)
        }))
        .await?;
    Ok(())
}
----

> Define your service as a parameter in you schema definition to reuse your policy with different services. The following code snippet provides an example schema definition for an external service:

Reviewer: This is not very accurate. The definition is a *must* to be able to make HTTP calls, not just so you can reuse it with different services.

[source,yaml]
----
properties:
      externalService:
          type: string
          format: service
      endpointPath:
          type: string
----

To learn more about defining parameters, see xref:policies-pdk-create-schema-definition.adoc[]

> In Rust, you can access the defined external service in the `Config` struct. To perform requests, call the HTTP client in your policy similar to the following example:

Reviewer: ```suggestion
In the policy Rust code, you can access the defined external service in the `Config` struct. To perform requests, call the HTTP client in your policy similar to the following example:
```

[source,rust]
----
let response = client
    .request(&config.external_service)
    .path(&config.endpoint_path)
    .headers(vec![("Content-Type", "application/json")])
    .body(r#"{"key": "value"}"#.as_bytes())
    .put().await?;
----

> To view an example policy that makes HTTP request, see the xref:policies-pdk-policy-templates.adoc[JWT Example Policy]. 

Reviewer: Why the JWT policy? This one does not make HTTP requests.

endif::[]
:imagesdir: ../assets/images

To interact with external services, the policy must use HTTP calls. You can only make HTTP calls from the request filter or the response filter. 

> To perform an HTTP call, either eject the client parameters into the `[entrypoint]` function or the wrapped functions:

Reviewer: ```suggestion
To perform an HTTP call, either inject the client parameters into the `[entrypoint]` function or the wrapper functions:
```

* Injecting the client parameters into the `[entrypoint]` function enables you to make calls in both the request and the response filter:
+
[source,Rust]
----
#[entrypoint]
async fn configure(launcher: Launcher, Configuration(bytes): Configuration, client: HttpClient) -> Result<()> {
    let config: Config = serde_json::from_slice(&bytes).unwrap();

    // Both request and the response handlers will use the client   
    let filter = on_request(|request_state| request_filter(request_state, &config, &client))
        .on_response(|response_state, request_data| {
            response_filter(response_state, request_data, &config, &client)
        });

    launcher.launch(filter).await?;
    Ok(())
}
----

> * Injecting the client parameters into one of the `on_request` or `on_response` wrapped functions enables you to only make calls in that function:

Reviewer: ```suggestion
* Injecting the client parameters into one of the `on_request` or `on_response` wrapper functions enables you to only make calls in that function:
```

To learn more about defining parameters, see xref:policies-pdk-create-schema-definition.adoc[]

> You must define your service as a parameter in you schema definition to make HTTP requests. To perform requests, call the HTTP client in your policy similar to the following example:

Reviewer: Why do we repeat this phrase again here? Here we would want to mention the `Config` struct

Author: Sorry about this one, I started editing a different section in last round.
